Mapgen enhancements

BevapDin opened the pull request:
This enhances / slightly restructures the mapgen code. Ignore the json changes they are cosmetic to match the changed format.

It gets rid of the enumerations, and uses specific classes instead.

For example the `jmapgen_place_special` class:

``` C++
struct jmapgen_place_special {
    jmapgen_int x;
    jmapgen_int y;
    jmapgen_place_special_op op;
    jmapgen_int amount;
    std::string type;
    jmapgen_place_special(jmapgen_int ix, jmapgen_int iy, jmapgen_place_special_op iop, jmapgen_int iamount, std::string itype
        ) : x(ix), y(iy), op(iop), amount(iamount), type(itype) {}
    void apply( map * m );
};
```

The `type` represents different things: for a toilet special it is ignored, for placing a NPC, it's its class, for placing a vending machine it's the item group of the items to put in it. Now there are separate classes for all the specials, each class contains only the member it requires (and they are named better):

``` C++
class jmapgen_vending_machine : public jmapgen_piece {
public:
    std::string item_group;
    jmapgen_vending_machine( JsonObject &jsi ) : jmapgen_piece()
    , item_group( jsi.get_string( "item_group", one_in( 2 ) ? "vending_food" : "vending_drink" ) )
    {
    }
    void apply( map &m, const size_t x, const size_t y, const float /*mon_density*/ ) const override {
        m.furn_set( x, y, f_null );
        m.place_vending( x, y, item_group );
    }
};
```

This also gets rid of the redundant code. Previously it would extract the json data into function variables, which would be forwarded to the constructor, which would forward them to the actual members. Adding / modifying a member required changing at least 3 places and additional the place where it got used. Now everything is done in the constructor, changing / adding a member requires only adding its declaration and adding its initialization.

It also makes the whole thing more self contained, that means adding a new mapgen thingy is easy: just copy an existing class like the above, add/change its members, add/change the initialization of the members and put the actual mapgen stuff into the `apply` function. I added the following to allow mapgen to place fields:

``` C++
class jmapgen_field : public jmapgen_piece {
public:
    field_id ftype;
    int density;
    int age;
    jmapgen_field( JsonObject &jsi ) : jmapgen_piece()
    , ftype( field_from_ident( jsi.get_string( "field" ) ) )
    , density( jsi.get_int( "density", 1 ) )
    , age( jsi.get_int( "age", 0 ) )
    {
        if( ftype == fd_null ) {
            jsi.throw_error( "invalid field type", "field" );
        }
    }
    void apply( map &m, const size_t x, const size_t y, const float /*mon_density*/ ) const override {
        m.add_field( point( x, y ), ftype, density, age );
    }
};
```

Those classes don't store a position on the map, that is now done separately. This allows mapping through the same letter array that is used to initialize the terrain/furniture:

``` JSON
"furniture": {
    "6": "f_table"
},
"items": {
    "6": { "item": "table_items" }
},
"fields": {
    "6": { "field": "fd_blood", "density": 2 }
}
```

The object with "table_items" has the same syntax as the objects in "place_groups" (it contains the item group id, and optionally a chance). In this case the x,y,repeat values are ignored (as the "6" already specifies where it is put).

In fact (nearly) all the things that can have a "place_foos" array and a "foos" object use the same syntax, only the entries in "place_foos" require the x,y coordinates. So you could write the above as

``` JSON
"place_items": [
    { "x": 10, "y": 10, "item": "table_items" }
],
"place_fields": [
    { "x": 10, "y": 10, "field": "fd_blood", "density": 2 }
]
```

I separated the "place_groups" and the "place_specials" arrays into several array:
- "place_items" places items from a group.
- "place_monsters" places monster spawn points.
- "place_vehicles" places vehicles.
- similar for all the specials.

Note that the entries for the specials don't need the type of the special listed separately:

``` JSON
"place_specials": [
    { "type": "toilet", "x": 6, "y": 22 }
],
"^^ this becomes:",
"place_toilets": [
    { "x": 6, "y": 22 }
],
"or alternatively:",
"toilets": {
    "g": { }
}
```

Note that the last entry has no further values, it could contain the amount of water that the toilet should have, but that value is optional.

If your mapgen image has the mapping "T" to `t_floor` and `f_bed`, you can now add a mapping to place items from the "bed" item group there:

``` JSON
"items": {
    "T": { "item": "bed", "chance": 80 }
}
```

This should fix #11057.

---

There are already many changes to the json data, but I would like to rename
- "place_items" to "place_item_groups" and its entries should contain a "group" entry, not an "item" entry.
- "place_monsters" to "place_monster_spawn", with an entry "group".
- the "vehicle" entry in "place_vehicles" should be named "type".

This would match their content better: the "item" entry in "place_items" does not refer to an item type, but to an item group, so it should be named "group".

Similar for monster groups, it should be named "place_monster_spawn" because it only adds _spawn points_, not actual monsters (the spawn point might not actually spawn a monster at all).

I also need to update the documentation when the changes here get approved and before it is merged.

bkentel commented:
Only enough time to take a peek at it now; it's a big change, but FWIW I like the direction.
Tiny suggestions, but maybe consider `override` for the derived `apply(...)` functions?

Frost-wood commented:
Looks good, it will definitely make adding buildings more easy and adding more mapgen thingy such as adding a single monster(like a boss, or a turret in a specific place) easier.

KrazyTheFox commented:
Oooh, I like this. I'm working on a not-so-super-secret project that will benefit from this greatly.

BevapDin commented:
@bkentel I added it to all the new classes, or did I forgot it somewhere? It's only add to new/changed code, there is still untouched code in mapgen.{cpp,h} that could use it, but that will be added separately.

Btw. if anyone is interested: the build fails because the json parser thought `JsonArray::size()` should return a signed int, which I changed into an unsigned int.

KrazyTheFox commented:
Any chance you could squeeze grass_or_dirt into this or another PR? That would be absolutely amazing.

BevapDin commented:
> Any chance you could squeeze grass_or_dirt into this or another PR? That would be absolutely amazing.

A second one.

KA101 commented:
Separate merge, I think.  Lotsa large infrastructure changes here.

KA101 commented on src/mapgen.cpp:
```diff
@@ -477,132 +477,252 @@ void mapgen_function_json::setup_setmap( JsonArray &parray ) {
+    }
+};
+/**
+ * Place a toilet with fuel in it.
```
gas pump, surely

vache commented:
I don't know if I'm missing something here, but why leave mon_density in jmapgen_piece::apply?  Are there any situations where its overridden in jmapgen_monster_group?  I've looked a little bit for it (at work so I can't really get too far into it), and I can't really find any examples.

BevapDin commented:
I kept it in as it was already there and I'm to lazy to change it. It is actually used in for example in my the necropoils, mall, DinoMod, Z-Level-Mod.
